drivers/crypto/crypto_stm32_sha.c

```diff
@@ -0,0 +1,221 @@
+#else
+#error No STM32 HW Hash Accelerator in device tree
+#endif
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_hash)
+#define STM32_HASH_TYPEDEF HASH_TypeDef
+#endif
+
+#define CRYPTO_HASH_MAX_SESSIONS CONFIG_CRYPTO_HASH_MAX_SESSIONS
+
+static struct crypto_stm32_sha_session crypto_stm32_sha_sessions[CRYPTO_HASH_MAX_SESSIONS];
+
+static int crypto_stm32_sha_get_unused_session_index(const struct device *dev)
+{
+	int i;
+	struct crypto_stm32_sha_data *data = CRYPTO_STM32_SHA_DATA(dev);
+
+	k_sem_take(&data->session_sem, K_FOREVER);
+
+	for (i = 0; i < CRYPTO_HASH_MAX_SESSIONS; i++) {
+		if (!crypto_stm32_sha_sessions[i].in_use) {
+			crypto_stm32_sha_sessions[i].in_use = true;
+			k_sem_give(&data->session_sem);
+			return i;
+		}
+	}
+
+	k_sem_give(&data->session_sem);
+	return -1;
+}
+
+static int stm32_hash_handler(struct hash_ctx *ctx, struct hash_pkt *pkt, bool finish)
+{
+	const struct device *dev = ctx->device;
+	struct crypto_stm32_sha_data *data = CRYPTO_STM32_SHA_DATA(dev);
+	struct crypto_stm32_sha_session *session = CRYPTO_STM32_SHA_SESSN(ctx);
+	HAL_StatusTypeDef status;
+
+	if (!pkt || !pkt->in_buf || !pkt->out_buf) {
+		LOG_ERR("Invalid packet buffers");
+		return -EINVAL;
+	}
+
+	if (!finish) {
+		LOG_ERR("Multipart hashing not supported yet");
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->device_sem, K_FOREVER);
+
+	switch (session->algo) {
+	case CRYPTO_HASH_ALGO_SHA224:
+		status = HAL_HASHEx_SHA224_Start(
+			&data->hhash,
+			pkt->in_buf,
+			pkt->in_len,
+			pkt->out_buf,
+			HAL_MAX_DELAY
+		);
+		break;
+	case CRYPTO_HASH_ALGO_SHA256:
+		status = HAL_HASHEx_SHA256_Start(
+			&data->hhash,
+			pkt->in_buf,
+			pkt->in_len,
+			pkt->out_buf,
+			HAL_MAX_DELAY
+		);
+		break;
+	default:
+		k_sem_give(&data->device_sem);
+		LOG_ERR("Unsupported algorithm in handler: %d", session->algo);
+		return -EINVAL;
+	}
+
+	k_sem_give(&data->device_sem);
+
+	if (status != HAL_OK) {
+		LOG_ERR("HAL HASH computation failed (status=%d)", status);
+		return -EIO;
+	}
+
+	LOG_DBG("Hash computation successful");
+	return 0;
+}
+
+static int stm32_hash_begin_session(const struct device *dev,
+				    struct hash_ctx *ctx,
+				    enum hash_algo algo)
+{
+	int ctx_idx;
+	struct crypto_stm32_sha_session *session;
+
+	ctx_idx = crypto_stm32_sha_get_unused_session_index(dev);
+	if (ctx_idx < 0) {
+		LOG_ERR("No free session for now");
+		return -ENOSPC;
+	}
+
+	session = &crypto_stm32_sha_sessions[ctx_idx];
+	memset(&session->config, 0, sizeof(session->config));
+	memset(session->digest, 0, sizeof(session->digest));
+
+	ctx->drv_sessn_state = session;
+
+	switch (algo) {
+	case CRYPTO_HASH_ALGO_SHA224:
+	case CRYPTO_HASH_ALGO_SHA256:
+		session->algo = algo;
+		break;
+	default:
+		LOG_ERR("Unsupported hash algorithm: %d", algo);
+		session->in_use = false;
+		return -EINVAL;
+	}
+
+	session->in_use = true;
+	ctx->hash_hndlr = stm32_hash_handler;
+	ctx->started = false;
+
+	LOG_DBG("begin_session (algo=%d)", algo);
+	return 0;
+}
```

Comment:
It would be wiser to delay "expensive" operations until we know that correct parameters have been provided:
```suggestion
static int stm32_hash_begin_session(const struct device *dev,
				    struct hash_ctx *ctx,
				    enum hash_algo algo)
{
	int ctx_idx;
	struct crypto_stm32_sha_session *session;

	ctx_idx = crypto_stm32_sha_get_unused_session_index(dev);
	if (ctx_idx < 0) {
		LOG_ERR("No free session for now");
		return -ENOSPC;
	}

	switch (algo) {
	case CRYPTO_HASH_ALGO_SHA224:
	case CRYPTO_HASH_ALGO_SHA256:
		break;
	default:
		LOG_ERR("Unsupported hash algorithm: %d", algo);
		return -EINVAL;
	}

	session = &crypto_stm32_sha_sessions[ctx_idx];
	memset(&session->config, 0, sizeof(session->config));
	memset(session->digest, 0, sizeof(session->digest));

	session->in_use = true;
	session->algo = algo;
	ctx->drv_sessn_state = session;
	ctx->hash_hndlr = stm32_hash_handler;
	ctx->started = false;

	LOG_DBG("begin_session (algo=%d)", algo);
	return 0;
}
```

Reply:
Done